### ISSUE TITLE
Make vars-on-top work with let and const

### ISSUE THREAD

**Issue by kentcdodds** [eslint/eslint#2865]
Right now this gives no warning:

```javascript
function foo(bar) {
  if(bar) {
    doSomething();
  }
  const baz = "bar";
  return baz;
}
foo();
```

given this config:

```json
{
  "rules": {
    "strict": 0,
    "vars-on-top": 2
  },
  "ecmaFeatures": {"blockBindings": true},
  "globals": {
    "doSomething": false
  }
}
```

(Note, it does warn if I include a regular `var` in there):

```javascript
function foo(bar) {
  if(bar) {
    doSomething();
  }
  const baz = "bar";
  var foobar = "eggs";
  return {baz: baz, foobar: foobar};
}
foo();
```

I would expect that it should warn in the first case.

**Comment by gyandeeps** [comment]
This was designed for `var` only because of hoisting. And hoisting doesn't apply to `let` or `const`.

**Comment by kentcdodds** [comment]
I'm pretty sure that it does... I believe it still hoists in the block... I may be mistaken.

**Comment by xjamundx** [comment]
I could see a new rule or option (purely for stylistic reasons) and no @kentcdodds it does not hoist!

**Comment by kentcdodds** [comment]
neat! I guess that babel doesn't do anything to cover that then because I get no transpiler warnings or interesting output to prevent it (see [this](http://babeljs.io/repl/#?experimental=true&evaluate=true&loose=false&spec=false&playground=true&code=function%20foo()%20%7B%0A%20%20const%20thing%20%3D%20'bar'%3B%0A%20%20if%20(thing%20%3D%3D%3D%20'bar')%20%7B%0A%20%20%20%20doSomething()%3B%0A%20%20%7D%0A%20%20console.log(baz)%3B%0A%20%20%0A%20%20const%20baz%20%3D%20'foo'%3B%0A%20%20return%20baz%20%2B%20thing%3B%0A%7D))

**Comment by xjamundx** [comment]
Oh you're right. Babel doesn't handle it yet:
http://kangax.github.io/compat-table/es6/#let_temporal_dead_zone_(strict_mode)

**Comment by kentcdodds** [comment]
Great. Thanks for explaining!

**Comment by kentcdodds** [comment]
Though, perhaps I should leave this open to add an option for stylistic reasons. Anyone think that would be handy? I'd definitely use it.

**Comment by nzakas** [comment]
@kentcdodds you're welcome to make a proposal. What is your expectation for how this rule should be modified to work?

**Comment by kentcdodds** [comment]
I've created a PR that adds tests to what I think should work. I may be able to look at actually implementing it sometime if you're cool with the suggestion.

**Comment by nzakas** [comment]
My concern with what you have proposed is that `let` and `const` are block-scoped, so it doesn't really make sense to force them to be declared at the top of a function. Maybe top of a block would make sense, but I'm unsure of that as well.

**Comment by kentcdodds** [comment]
Yes, good point. Would you like me to add tests for that as well?

**Comment by nzakas** [comment]
No, we aren't going to merge failing tests, so they're of little use.

Do others agree making such a change makes sense?

**Comment by gyandeeps** [comment]
I think if we want to consider `let` and `const` then it should be an option and by default off. Because everyone has their own views when it comes to block scoped declaration and TDZ mixup.

**Comment by kentcdodds** [comment]
I didn't intend for that PR to get merged before the implemenation. More of an actual example of what I'm suggesting :-)

I agree with @gyandeeps that it should be an option that is off by default because it's more stylistic than anything.

**Comment by nzakas** [comment]
@kentcdodds it's more useful to paste examples into the issue.

@gyandeeps agreed

**Comment by gyandeeps** [comment]
@nzakas Based on the convo on top I think we can add accepted tag here?

**Comment by IanVS** [comment]
:-1: This seems like a pretty strange rule to put in core.   Mixing `let` with `var`, which have very different behavior, will only lead to confusion in my opinion.  If for some reason your style is to put them on top, and you want to enforce that convention, I'd argue for creating a plugin.

**Comment by kentcdodds** [comment]
I'm fine with that.

- Kent C. Dodds

(Sent from my mobile device, please forgive typos or brevity)

**Comment by nzakas** [comment]
Okay, seems like consensus on a plugin, so closing. @kentcdodds if you go the plugin route, please paste a link to your plugin here so others can benefit.

**Comment by localpcguy** [comment]
Was a plugin for this ever made? I agree that it should be block scoped, and fine with it being off by default, but it'd be nice to be able to enforce this from a stylistic aspect.

**Comment by tdmalone** [comment]
@kentcdodds Just pinging you to check if you did ever manage to make a plugin for this.

**Comment by kentcdodds** [comment]
Nope, I never worked on it. I lost interest in it.
